login/pam.c

```diff
@@ -0,0 +1,272 @@
+// Copyright 2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <errno.h>
+#include <limits.h>
+#include <stdlib.h>
+#include <string.h>
+#include <syslog.h>
+#include <time.h>
+
+#include <security/pam_ext.h>
+#include <security/pam_modules.h>
+
+#include "base64.h"
+#include "config.h"
+#include "lockdown.h"
+#include "login.h"
+#include "ui.h"
+
+#define MODULE_NAME "pam_glome"
+
+static int parse_pam_args(pam_handle_t *pamh, int argc, const char **argv,
+                          login_config_t *config) {
+  memset(config, 0, sizeof(login_config_t));
+  int errors = 0;
+  char *endptr;
+  long l;
+
+  // Setting defaults.
+  config->reboot_user = DEFAULT_REBOOT_USER;
+  config->login_path = DEFAULT_LOGIN_PATH;
+  config->lockdown_path = NULL;
+  config->url_prefix = NULL;
+  config->auth_delay_sec = DEFAULT_AUTH_DELAY;
+  config->input_timeout_sec = DEFAULT_INPUT_TIMEOUT;
+  config->options = SYSLOG;
```

Comment:
Some of these options are imho not relevant for the PAM module:

- `auth_delay_sec` should probably be implemented with `pam_faildelay`
- Not entirely sure about `input_timeout_sec`, but I did not find other PAM modules with a configurable timeout.
- Lockdown path is a runtime environment detail that should be handled outside of the GLOME module (c.f. `pam_time`).
- The PAM module arguably should not cause reboots, so we don't need to know about a `reboot_user`.
- The same goes for `login_path`.

Actually, I think it might be better not to use the configuration facilities of `/sbin/glome-login` at all and rather implement only the strictly necessary options (service key, debug, machineid?) as PAM module arguments.

Reply:
Removed the parameters that are not relevant for PAM module.

As for `input_timeout_sec` it seems that PAM prompt waits 60 seconds for input so this should work by default too.